os/pm/pm_sleep.c

```diff
@@ -92,50 +92,52 @@ static void pm_timer_callback(pm_timer_t* timer)
  *      3. NORMAL to SLEEP state threshold time is large
  * 
  * Parameters:
- *   timer_interval - expected board sleep duration (in millisecond)
+ *   milliseconds - expected board sleep duration
  *
  * Return Value:
- *   0 - success
- *   -1 - error
+ *   OK - success
+ *   ERROR - error
  *
  ************************************************************************/
 
 int pm_sleep(int milliseconds)
 {
-        int timer_interval;
-
-        /* Converting from milliseconds to ticks */
-	timer_interval = MSEC2TICK(milliseconds);
-
-        pm_timer_t *timer = pm_timer_create();
-        irqstate_t state;
-        if (timer == NULL) {
-                pmdbg("Unable to create pm timer\n");
-                return PM_TIMER_FAIL;
-        }
-
-        /* initialize the timer's semaphore. It will be used to lock the
-         * thread before sleep and unlock after expire */
-        sem_init(&timer->pm_sem, 0, 0);
-
-        timer->callback = pm_timer_callback;
-
-        /* Now add the timer in the list 
-         * Adding a wakeup timer in the linked list should be atomic.
-         * Otherwise there is a chance of wrong ordering of the list.*/
-        state = enter_critical_section();
-
-        /* Add the timer in the g_pm_timer_activelist */
-        timer->delay = timer_interval;
-        pm_timer_add(timer);
-
-        leave_critical_section(state);
-
-        /* sem_wait untill the timer expires */      
-        sem_wait(&timer->pm_sem);
-
-        /* When the semaphore is freed, make the pm timer free */
-	pm_timer_delete(timer);
-
-        return PM_TIMER_SUCCESS;
+	int ret;
+	WDOG_ID wdog;
+	sem_t pm_sem;
+
+	/* initialize the timer's semaphore. It will be used to lock the
+	 * thread before sleep and unlock after expire */
+	sem_init(&pm_sem, 0, 0);
+
+	wdog = wd_create();
+	if (!wdog) {
+		set_errno(EAGAIN);
+		return ERROR;
+	}
+
+	if (wd_setwakeupsource(wdog) != OK) {
+		wd_delete(wdog);
+		return ERROR;
+	}
+
+	wdparm_t wdparm;
+	wdparm.pvarg = &pm_sem;
+	ret = wd_start(wdog, MSEC2TICK(milliseconds), (wdentry_t)pm_timer_callback, 1, wdparm);
```

Comment:
Please change `wdparm` to `wdparm.pvarg`

Reply:
Thank you for review, I changed